### PR TITLE
Initial pass at single node recombination

### LINKED ISSUE #1974: Rename or change behaviour of NODE_IS_CA_EVENT
At the moment, when we set `record_full_arg=True`, we mark those nodes that are "Common Ancestor" events but *not* coalescence events with `NODE_IS_CA_EVENT`. In fact, the CA class should also include ("normal") coalescent nodes too, as all nodes should be either CA or RE nodes (or both), so we should either change the flag name or the behaviour.

This can probably be done as part of #1942

### PR DESCRIPTION
**Pull request by JereKoskela**
This changes the infrastructure for recombination nodes under the `full-arg` from two nodes per recombination to one. I've introduced a new flag for the new recombination node, but commented out the old one rather than introduced error handling if old `MSP_NODE_IS_RE_EVENT`s are encountered. Is that something you can add easily along with the breakpoint metadata @jeromekelleher? Background for this change is in #1942.

As discussed, the likelihood function will be broken until I've had a chance to fix it, which I'll be able to do one #1979 has been resolved.

I also changed the flag for `MSP_NODE_IS_CA_EVENT` to the more descriptive `MSP_NODE_IS_NONGENETIC_CA`. I'm happy with further tweaks if anyone has better ideas.

To my surprise, this change causes `test_multimerger` in `test_ancestry.py` to fail. Apparently, a full-arg has 172 edges while its simplified tree sequence has 173. The test involves an instantaneous bottleneck, which is a part of the code I haven't changed, and with which I'm not too familiar. @hyanwong, based on your visualisation plots, do you think there can be some edge case in which simplifying increases the number of edges?

Closes #1974

### CODE REVIEW

**Comment by hyanwong** [comment]
> To my surprise, this change causes `test_multimerger` in `test_ancestry.py` to fail. Apparently, a full-arg has 172 edges while its simplified tree sequence has 173. The test involves an instantaneous bottleneck, which is a part of the code I haven't changed, and with which I'm not too familiar. @hyanwong, based on your visualisation plots, do you think there can be some edge case in which simplifying increases the number of edges?

Yes, definitely. This can be quite common. It's exactly the same issue we have been debating over in https://github.com/tskit-dev/what-is-an-arg-paper/issues/17#issuecomment-1009410925

**Comment by JereKoskela** [comment]
> > @hyanwong, based on your visualisation plots, do you think there can be some edge case in which simplifying increases the number of edges?
> 
> Yes, definitely. This can be quite common. It's exactly the same issue we have been debating over in [tskit-dev/what-is-an-arg-paper#17 (comment)](https://github.com/tskit-dev/what-is-an-arg-paper/issues/17#issuecomment-1009410925)

Excellent, thanks. In that case I'll go ahead and just delete that assert from the test.

File: lib/likelihood.c [hunk 1]
```diff
@@ -121,12 +121,13 @@ msp_log_likelihood_arg(tsk_treeseq_t *ts, double r, double Ne, double *r_lik)
     double material = lineages * tsk_treeseq_get_sequence_length(ts);
     double material_in_children, material_in_parent, rate, gap;
     double lik = 0;
+    double left_end, right_end;
```
Comment (jeromekelleher) [inline lib/likelihood.c:124]: Did you intend to include these changes in the likelihood calculation? I thought the idea was to leave this to a follow-up?

File: lib/msprime.c [hunk 2]
```diff
@@ -5090,9 +5084,9 @@ msp_run(msp_t *self, double max_time, unsigned long max_events)
     }
     if (self->store_full_arg
         && !(self->model.type == MSP_MODEL_HUDSON || self->model.type == MSP_MODEL_SMC
-               || self->model.type == MSP_MODEL_SMC_PRIME
```
Comment (jeromekelleher) [inline lib/msprime.c:5093]: Looks like you're on the wrong version of clang-format. See [here](https://tskit.dev/tskit/docs/stable/development.html#sec-development-c-requirements) for info on getting the right one.

Yes, clang-format is awful for changing the format between versions in a non-compatible way.

**Review by jeromekelleher — COMMENTED** [review]
Basic changes LGTM.

I'm not sure what the simplest approach to dealing with the likelihood code is?

File: lib/msprime.c (hunk 2 above)
Comment (JereKoskela) [reply lib/msprime.c]: Thanks! I think that's fixed now.

File: lib/likelihood.c (hunk 1 above)
Comment (JereKoskela) [reply lib/likelihood.c:124]: Yes. This works whenever there isn't an offending diamond in the ARG, and is as far as I got before realising the problem. I thought it would be quickest to include it here to have as a starting point for the fix once the metadata is in. Happy to do something different too if you think that's not good practice.

Comment (jeromekelleher) [reply lib/likelihood.c:124]: I think it probably is a good idea to back out of these changes on this branch, and try to keep them separate.

The easiest thing to do is to start a new branch based on your current one, which can be the starting point for likelihood updates. You can then reset the state of the likelihood files to what they are in upstream/main on this branch and it should all go smoothly.

Comment (JereKoskela) [reply lib/likelihood.c:124]: Sounds good to me. I'll push one more update to that effect and ping you.

**Comment by JereKoskela** [comment]
Ok @jeromekelleher, I've reverted both the C and Python implementations of the likelihoods back to the current versions. The remaining changes in the tests files are either updating NODE_IS_RE_EVENT labels to NODE_IS_RECOMBINANT, or changing the manual construction of the example ARGs there to conform to the new data structure.

The test pipeline will obviously fail because the likelihood functions are wrong. I'm not sure what the best way around that is.

**Comment by jeromekelleher** [comment]
I've added a commit to skip the tests temporarily @JereKoskela. The other failures are unrelated I think, I'll fix separately and update this PR afterwards.

Look OK to you?

**Comment by jeromekelleher** [comment]
@mergifyio rebase

**Review by jeromekelleher — APPROVED** [review]
LGTM - let's merge an move on
